### PR TITLE
feat(project): optionally parse error response per request, endpoint …

### PR DESCRIPTION
**Pull request by doktordirk**
…or globally

closes https://github.com/SpoonX/aurelia-api/issues/134

---
This change is [Reviewable](https://reviewable.io/reviews/spoonx/aurelia-api/150)

### CODE REVIEW

**Comment by doktordirk** [comment]
msalman86 have a look

File: src/config.js [hunk 1]
```diff
@@ -26,11 +39,15 @@ export class Config {
...
    * @see http://aurelia.io/docs.html#/aurelia/fetch-client/latest/doc/api/class/HttpClientConfiguration
    * @return {Config}
+   * @chainable
    */
   registerEndpoint(name, configureMethod, defaults) {
     let newClient        = new HttpClient();
     this.endpoints[name] = new Rest(newClient, name);
 
+    // get global defaults
+    if (this.defaults) defaults = extend(true, {}, this.defaults, defaults);
```
Comment (RWOverdijk) [inline src/config.js]: Curly braces

File: src/rest.js [hunk 2]
```diff
@@ -37,6 +44,10 @@ export class Rest {
   request(method, path, body, options = {}) {
     let requestOptions = extend(true, {headers: {}}, this.defaults, options, {method, body});
 
+    // extract parseError option
+    let parseError = requestOptions.parseError;
+    delete requestOptions.parseError;
```
Comment (RWOverdijk) [inline src/rest.js]: Newline

File: src/aurelia-api.js [hunk 3]
```diff
@@ -4,4 +4,6 @@ export function configure(aurelia, configCallback) {
   let config = aurelia.container.get(Config);
 
   configCallback(config);
+
+  config.configured = true;
```
Comment (RWOverdijk) [inline src/aurelia-api.js]: What's this?

File: src/config.js [hunk 4]
```diff
@@ -13,11 +14,23 @@ export class Config {
...
 
   /**
+   * Fetch request defaults applied to all endpoints
+   * @param {{}} defaults The defaults opject
+   */
+  defaults = null;
+
+  /**
+   * The plugin's configured status. Set on plugin('aurelia-api', ...)
+   * @param {[Boolean]} configured The plugin's configured status
```
Comment (RWOverdijk) [inline src/config.js]: Newline

Comment (RWOverdijk) [inline src/config.js]: What does {[Boolean]} mean? Afaik that's not the type you want.

File: src/config.js [hunk 5]
```diff
@@ -13,11 +14,23 @@ export class Config {
...
   /**
    * Current default endpoint if set
-   * @param {[Rest]} Default Rest client
+   * @param {[Rest]} defaultEndpoint The Rest client
    */
   defaultEndpoint = null;
 
   /**
+   * Fetch request defaults applied to all endpoints
+   * @param {{}} defaults The defaults opject
```
Comment (RWOverdijk) [inline src/config.js]: Newline

File: src/rest.js [hunk 6]
```diff
@@ -6,11 +6,18 @@ import extend from 'extend';
  */
 export class Rest {
 
+  /**
+   * Fetch options for all requests
+   * See also https://developer.mozilla.org/en-US/docs/Web/API/Request/Request
+   * Accepts additional option parseError, which will be removed before fetching
+   * @param {{}} defaults The defaults object
```
Comment (RWOverdijk) [inline src/rest.js]: Newline

File: src/rest.js [hunk 7]
```diff
@@ -50,7 +61,11 @@ export class Rest {
         return response.json().catch(error => null);
       }
 
-      throw response;
+      if (parseError) {
+        return response.json().catch(() => response).then(error => Promise.reject(error));
```
Comment (RWOverdijk) [inline src/rest.js]: Why are you returning the response as a rejected promise, in stead of just using the catch?

File: doc/api_rest.md [hunk 8]
```diff
@@ -26,10 +26,11 @@ import {Rest} from 'aurelia-api';
 
 All methods will:
 
-* maintain trailing slashes of the resource parameter 
 * stringify the body, if it is an object and the `Content-Type` is set to `application/json` (the default).
 * convert the body to querystring format, if the body is an object and the `Content-Type` is set to any other value.
 * leave the body unchanged, if the `Content-Type` is not set or when the body is not an object.
+* return the server on error the response JSON parsed if possible and parseError is set true in the defaults or the options
```
Comment (RWOverdijk) [inline doc/api_rest.md]: This sentence doesn't make a lot of sense. You return a server?

File: src/config.js [hunk 9]
```diff
@@ -13,11 +14,23 @@ export class Config {
 
   /**
    * Current default endpoint if set
-   * @param {[Rest]} Default Rest client
+   * @param {[Rest]} defaultEndpoint The Rest client
```
Comment (RWOverdijk) [inline src/config.js]: This is not a valid type.

File: src/config.js [hunk 10]
```diff
@@ -26,11 +39,15 @@ export class Config {
...
    */
   registerEndpoint(name, configureMethod, defaults) {
     let newClient        = new HttpClient();
     this.endpoints[name] = new Rest(newClient, name);
 
+    // get global defaults
+    if (this.defaults) defaults = extend(true, {}, this.defaults, defaults);
+
     // set custom defaults to Rest
     if (defaults !== undefined) this.endpoints[name].defaults = defaults;
```
Comment (RWOverdijk) [inline src/config.js]: Curly braces

File: src/rest.js (hunk 7 above)
Comment (doktordirk) [inline src/rest.js]: the response might not be a json. hence catching it like aver in line 61. so error will be either the response or JSON.parse(response) there. 

i used Promise.reject due to an error in thinking. what i meant is than one should throw resp. reject with type Error. that's now tricky to do BC unless we add an option for that

Comment (doktordirk) [inline src/rest.js]: hmm, what'e nicer
then(Promise.reject);
or
then(error => throw error);

File: src/aurelia-api.js (hunk 3 above)
Comment (doktordirk) [inline src/aurelia-api.js]: i guess you figured it out.

the reason is, that changing to parsed response can mess up plugins which expect unparsed error. so i wanted to add a simple check that one doesn't change to parseError=true globally eg within a plugin

File: src/rest.js (hunk 7 above)
Comment (RWOverdijk) [inline src/rest.js]: Or neither? It will already return a rejected promise. You can leave off the catch.

**Comment by doktordirk** [comment]
@RWOverdijk i still needed to reject, but you're right. JSON.parse error already had been covered

fyi: there is one additional fix in there for body === null

**Comment by RWOverdijk** [comment]
Can't it be implemented by users? It's an option you can pass in, it feels oddly specific to implement.

**Comment by doktordirk** [comment]
i guess an interceptor could do it as well

**Pull request closed without merging**